storage/connect/tabrest.cpp

```diff
@@ -353,12 +186,140 @@ PTDB RESTDEF::GetTable(PGLOBAL g, MODE m)
   if (trace(515))
     htrc("REST GetTable mode=%d\n", m);
 
-  if (m != MODE_READ && m != MODE_READX && m != MODE_ANY) {
-    strcpy(g->Message, "REST tables are currently read only");
+  if (m != MODE_READ && m != MODE_READX && m != MODE_ANY)
+  {
+    snprintf(g->Message, sizeof(g->Message), "REST tables are currently read only");
     return NULL;
-  } // endif m
+  }
 
   return Tdp->GetTable(g, m); // Leave file type do the job
 } // end of GetTable
 
+
+/***********************************************************************/
+/*  curl_init: Initilize curl                                          */
+/***********************************************************************/
+int RESTDEF::curl_init(PGLOBAL g)
+{
+  CURLcode curl_res = curl_global_init(CURL_GLOBAL_ALL);
+  if (curl_res != CURLE_OK)
+  {
+    if (curl_res)
+    {
+      char msg[512];
+      snprintf(msg, 512, "unable to initialize curl library, "
+                         "curl returned this error code: %u "
+                         "with the following error message: %s",
+                         curl_res, curl_easy_strerror(curl_res));
+      snprintf(g->Message, sizeof(g->Message), "%s", msg);
+      return 1;
+    }
+  }
+  curl_inited = true;
+  return 0;
+}
+
+
+/***********************************************************************/
+/*  curl_deinit: Cleanup curl                                          */
+/***********************************************************************/
+void RESTDEF::curl_deinit()
+{
+  if (curl_inited)
+  {
+    curl_global_cleanup();
+    curl_inited = false;
+  }
+}
+
+
+/***********************************************************************/
+/*  WriteMemoryCallback: Curl callback function                        */
+/***********************************************************************/
+static size_t WriteMemoryCallback(void *contents,
+                                  size_t size __attribute__((unused)),
+                                  size_t nmemb, void *userp)
+{
+  struct MemoryStruct *mem = (struct MemoryStruct *)userp;
+  char *ptr = (char *)realloc(mem->memory, mem->size + nmemb + 1);
+  if (ptr == NULL)
+      return 0;
+  mem->memory = ptr;
+  memcpy(&(mem->memory[mem->size]), contents, nmemb);
+  mem->size += nmemb;
+  mem->memory[mem->size] = 0;
+  return nmemb;
+}
+
+
+/***********************************************************************/
+/*  curl_run: Retrieve the REST answer by executing cURL.              */
+/***********************************************************************/
+int RESTDEF::curl_run(PGLOBAL g)
+{
+  CURL *curl = curl_easy_init();
+  CURLcode curl_res = CURLE_OK;
+  char buf[512];
+  long http_code = 0;
+  char curl_errbuf[CURL_ERROR_SIZE];
+  if (!curl)
+  {
+    snprintf(g->Message, sizeof(g->Message), "Cannot initilize curl session.");
+    return 1;
+  }
+  curl_errbuf[0] = '\0';
+  if (Uri)
+  {
+    if (*Uri == '/' || Http[strlen(Http) - 1] == '/')
```

Comment:
This looks really odd. What's it doing?

Reply:
This is inherited from old code, about parsing the argument from table options.